docs/concepts/function-components/pre-defined-hooks.md

```diff
@@ -1,27 +1,322 @@
 ---
-title: Pre-defined hooks
-description: The pre-defined hooks that Yew comes with 
+title: Pre-defined Hooks
+description: The pre-defined Hooks that Yew comes with 
 ---
 
-:::important contribute
-Contribute to our docs: Add details about pre-defined hooks
+:::note Why do Hooks return `Rc`?
 
-Take a look at the [project board](https://github.com/yewstack/yew/projects/3) for details and consider helping out.
+In most cases, you'll be cloning the values returned from the Hooks.
+As it is generally expensive to clone such values, they're `Rc`ed, so they can be cloned relatively cheaply.
+
+The following example shows one of the most common cases which requires cloning the values:
+
+```rust
+let (text, set_text) = use_state(|| "Hello".to_owned());
+let onclick = {
+    let text = Rc::clone(&text);
+    // Values must be moved into this closure so in order to use them later on, they must be cloned
+    Callback::from(move |_| set_text(format!("{} World", text))) 
+};
+
+// If `text` wasn't cloned above, it would've been impossible to use it here
+html! { text }
+```
 :::
 
 ## `use_state`
 
+`use_state` is used to mange state in a function component.
+It returns a `Rc` pointing to the value of the hook's state, and a setter function.
+
+The hook takes a function as input which determines the initial state.
+This value remains up-to-date on subsequent renders.
+
+The setter function is used to update the value and trigger a re-render.
+
+### Example
+
+```rust
+#[function_component(UseState)]
+fn state() -> Html {
+    let (
+        counter, // the returned state
+        set_counter // setter to update the state
+    ) = use_state(|| 0);
+    let onclick = {
+        let counter = Rc::clone(&counter);
+        Callback::from(move |_| set_counter(*counter + 1))
+    };
+
+    html! {
+        <div>
+            <button onclick=onclick>{ "Increment value" }</button>
+            <p>
+                <b>{ "Current value: " }</b>
+                { counter }
+            </p>
+        </div>
+    }
+}
+```
 
 ## `use_ref`
+`use_ref` is used for obtaining a mutable reference to a value.
+Its state persists across renders.
+
+It is important to note that you do not get notified of state changes.
+If you need the component to be re-rendered on state change, consider using [`use_state`](#use_state).
+
+### Example
+
+```rust
+#[function_component(UseRef)]
+fn ref_hook() -> Html {
+    let (outer_html, set_outer_html) = use_state(String::new);
+
+    let node_ref = use_ref(NodeRef::default);
+
+    let onclick = {
+        let node_ref = Rc::clone(&node_ref);
+
+        Callback::from(move |e| {
+            let outer_html = (*node_ref.borrow().deref())
+                .cast::<yew::web_sys::Element>()
+                .unwrap()
+                .outer_html();
+            set_outer_html(outer_html)
+        })
+    };
+    html! {
+        <>
+            <div id="result" ref=(*node_ref.borrow_mut().deref_mut()).clone()>{ "Filler" }</div>
```

Comment:
That is a good point. Looking though some examples of React's `useRef` online, I found [this example](https://gist.github.com/reireynoso/f373a0ad80f6a455cb8fa0089c256e9a) from [here](https://medium.com/javascript-in-plain-english/implementing-useref-in-react-732908aa1998). I think we should adopt that example since it does a better job at showcasing how `use_ref` can be used.